### PR TITLE
Entailment model

### PR DESCRIPTION
**Pull request by vyskoto4**

### CODE REVIEW

File: data/rte/snli/README.md [hunk 1]
```diff
@@ -34,6 +34,8 @@ Note that NO DROPOUT is applied for any of the models.
 |                          |±0.037068 |±0.012773 |±0.008269 |
 | attn1511                 | 0.828611 | 0.781472 | 0.773921 | ``dropout=0`` ``inp_e_dropout=0``
 |                          |±0.013870 |±0.002253 |±0.004134 |
+| cnn                      | 0.837948 | 0.781701 | 0.776135 | ``inp_w_dropout=0`` ``dropout=0`` ``inp_e_dropout=0``
+|                          |±0.008889 |±0.004660 |±0.003057 |
```
Comment (pasky) [inline data/rte/snli/README.md:38]: If this is a plain cnn, can you please sort it between rnn and rnncnn as in other task tables?

Also, we should re-benchmark everything after merging with bigvocab as the table is out-of-date. That's not a problem for merging this in, but for now we should at least add a comment in front of the table warning about this.

File: models/rnn.py [hunk 2]
```diff
@@ -54,18 +54,7 @@ def prep_model(model, N, s0pad, s1pad, c):
...
-                                          W_regularizer=l2(c['l2reg']), activation=c['pact']))
-        # This dropout is controversial; it might be harmful to apply,
-        # or at least isn't a clear win.
-        # model.add_shared_node(name='projdrop', inputs=['e0p', 'e1p'], outputs=['e0p_', 'e1p_'],
-        #                       layer=Dropout(c['dropout'], input_shape=(N,)))
-        # return ('e0p_', 'e1p_')
-        return ('e0p', 'e1p')
-    else:
-        return ('e0s_', 'e1s_')
+    # Projection deleted
```
Comment (pasky) [inline models/rnn.py:59]: I guess we shouldn't merge this, right? :) Can't we just set project=False if it's not appropriate for some model or task/model combination?

File: models/rnn.py [hunk 3]
```diff
@@ -54,18 +54,7 @@ def prep_model(model, N, s0pad, s1pad, c):
                 dropout=c['dropout'], dropoutfix_inp=c['dropoutfix_inp'], dropoutfix_rec=c['dropoutfix_rec'],
                 sdim=c['sdim'],
                 rnnbidi=c['rnnbidi'], rnn=c['rnn'], rnnact=c['rnnact'], rnninit=c['rnninit'],
-                rnnbidi_mode=c['rnnbidi_mode'], rnnlevels=c['rnnlevels'])
+                rnnbidi_mode=c['rnnbidi_mode'], rnnlevels=c['rnnlevels'],return_sequences=True)
```
Comment (pasky) [inline models/rnn.py:57]: Why?

File: pysts/eval.py [hunk 4]
```diff
@@ -46,11 +46,11 @@ def multiclass_accuracy(y, ypred):
...
-    result = np.zeros(ypred.shape)
+    result = np.zeros(y.shape)
     clss=y.shape[1]
     class_correct=np.zeros(clss)
     ok=0
-    for row in range(ypred.shape[0]):
+    for row in range(y.shape[0]):
         result[row,np.argmax(ypred[row])]=1
         for cls in range(clss):
             if y[row,cls]==result[row,cls]:
```
Comment (pasky) [inline pysts/eval.py:56]: Not a problem to merge this (I think!), but do you remember the rationale for changing this?

File: pysts/eval.py [hunk 5]
```diff
@@ -261,7 +261,7 @@ def eval_ubuntu(ypred, s0, y, name):
 def eval_rte(ypred, y, name):
     cls_names = ['contradiction', 'neutral', 'entailment']
     rawacc, cls_acc = multiclass_accuracy(y, ypred)
-    print('%s Accuracy: %.3f, %s accuracy %.3f, %s accuracy %.3f, %s accuracy %.3f' % (name, rawacc,
+    print('%s Accuracy: %.3f, %s accuracy %.3f, %s accuracy %.3f, %s accuracy %.3f' % (name, rawacc,                                                                                        
                                                                                         cls_names[0], cls_acc[0],
                                                                                         cls_names[1], cls_acc[1],
                                                                                         cls_names[2], cls_acc[2]))
```
Comment (pasky) [inline pysts/eval.py:267]: Probably not something for us to merge?

File: tasks/__init__.py [hunk 6]
```diff
@@ -19,7 +19,7 @@ def default_config(model_config, task_config):
     c['embdim'] = 300
     c['embprune'] = 100
     c['embicase'] = False
-    c['inp_e_dropout'] = 1/2
+    c['inp_e_dropout'] = 0
     c['inp_w_dropout'] = 0
     c['e_add_flags'] = True
```
Comment (pasky) [inline tasks/__init__.py:25]: Not something to merge? (If you need to override this for SNLI, just do it there.)

File: tasks/__init__.py [hunk 7]
```diff
@@ -34,7 +34,7 @@ def default_config(model_config, task_config):
     c['opt'] = 'adam'
     c['fix_layers'] = []  # mainly useful for transfer learning, or 'emb' to fix embeddings
     c['batch_size'] = 160
-    c['nb_epoch'] = 16
+    c['nb_epoch'] = 1
     c['nb_runs'] = 1
     c['epoch_fract'] = 1
```
Comment (pasky) [inline tasks/__init__.py:40]: Not something to merge.

File: tasks/rte.py [hunk 8]
```diff
@@ -34,6 +34,7 @@ def __init__(self):
         self.s1pad = self.spad
         self.emb = None
         self.vocab = None
+	self.add_output_layers=False
 
     def config(self, c):
         c['loss'] = 'categorical_crossentropy'
```
Comment (pasky) [inline tasks/rte.py:40]: Whitespace trouble? (tab instead of spaces ...and no spaces around = though that's nitpicking.)

File: tasks/rte.py [hunk 9]
```diff
@@ -77,16 +78,18 @@ def prep_model(self, module_prep_model):
...
+                	       	layer=Activation('sigmoid'))
+        	model.add_node(name='scoreS1', input=ptscorer(model, final_outputs, self.c['Ddim'], N, self.c['l2reg'], pfx="out1", **kwargs),
+                	      	layer=Activation('sigmoid'))
+        	model.add_node(name='scoreS2', input=ptscorer(model, final_outputs, self.c['Ddim'], N, self.c['l2reg'], pfx="out2", **kwargs),
+                		layer=Activation('sigmoid'))
+		model.add_node(name='scoreV', inputs=['scoreS0', 'scoreS1', 'scoreS2'], merge_mode='concat', layer=Activation('softmax'))
+                model.add_output(name='score', input='scoreV')
+	else:
+		model.add_output(name='score',input=final_outputs[0])
+	return model
```
Comment (pasky) [inline tasks/rte.py:92]: Strange indentation here. (And space after , missing, sorry for nitpicknig.)

File: tasks/rte.py [hunk 10]
```diff
@@ -100,10 +103,11 @@ def build_model(self, module_prep_model, do_compile=True):
 
         if do_compile:
             model.compile(loss={'score': self.c['loss']}, optimizer=self.c['opt'])
+	model.summary()
```
Comment (pasky) [inline tasks/rte.py:106]: indent (and do we want to merge this? maybe, just checking)

File: tasks/rte.py [hunk 11]
```diff
@@ -113,7 +117,10 @@ def eval(self, model):
             if gr is None:
                 res.append(None)
                 continue
-            ypred = model.predict(gr)['score']
+	    ypred=[]
```
Comment (pasky) [inline tasks/rte.py:120]: indent

File: tasks/snli.py [hunk 12]
```diff
@@ -31,6 +31,7 @@
 
 class SnliTask(RTETask):
     def __init__(self):
+	super(self.__class__, self).__init__()
```
Comment (pasky) [inline tasks/snli.py:34]: indent

File: models/rnn.py (hunk 3 above)
Comment (vyskoto4) [inline models/rnn.py:57]: We do not want to merge this.

File: models/rnn.py (hunk 2 above)
Comment (vyskoto4) [inline models/rnn.py:59]: Again. Sorry. We do not want to merge this.

**Comment by pasky** [comment]
In commit 16cfa46ed705bd33d2e942f871558bca3b7b686e "Some old changes" you are reverting all your architectural changes in the SNLI task. Is that intentional?

It is maybe a bit of a mess, with several changes tangled together and that may result in this confusion, so if it's not too much of a hassle, what I'd like to propose is to build two separate PR's from scratch (not based on the commit history of this PR):
- Porting RTE/SNLI to f/bigvocab.  That should be mostly a transposition of the current final diff to just a few new commits, right?
- Adding new attention-based architecture.  We can start iterating this by just having the change in the SNLI task file as in the historical commits on this branch, but I'd like to eventually merge a model that implements this (and can be then tried on other tasks).   But I can do part of that work myself too if you wish, I'd just need a good starting point diff to evolve it to a good form.

Does this all make sense?

**Comment by vyskoto4** [comment]
I agree. It will be definitely easier for you. I'll prepare the two
branches. Actually they are ready now but I just want to test if it works.

út 19. 7. 2016 v 17:58 odesílatel Petr Baudis notifications@github.com
napsal:

**Comment by pasky** [comment]
This PR got auto-closed since it's against the f/bigvocab branch which got deleted (actually, merged to master!).

However, @vyskoto4, you mentioned 23 days ago that "the branches are ready now" - can we make the second phase a reality, please? :) No need for overnight rush but I'd be so happy to see that PR open maybe during next week?

**Pull request closed without merging**